### ISSUE TITLE
`unused_assignments` is fired when it shouldn't

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust#105272]
Feel free to close if duplicated or if the behavior is intended.

If a variable is reassigned to another variable, then the lint will think that the original variable was never read even if the reassigned variable is being used.

```rust
enum Foo {
    Bar(i32),
    Baz
}

fn foo(foo: Foo) {
    // "Higher" variable because of lifetimes
    let mut placeholder = String::new();
    let _str = match foo {
        // The contents of `placeholder` are being assigned to `_str`
        Foo::Bar(num) => {
            let num_string = format!("{}", num);
            placeholder = num_string;
            placeholder.as_str()
        }
        Foo::Baz => {
            "Nothing"
        }
    };
    // `_str` is read and so `placeholder` but the lint fires anyway
    println!("{}", _str);
}

fn main() {
    foo(Foo::Bar(1));
    foo(Foo::Baz);
}
```

**Comment by jruderman** [comment]
The warning is:

```
warning: value assigned to `placeholder` is never read
 --> src/main.rs:8:13
  |
8 |     let mut placeholder = String::new();
  |             ^^^^^^^^^^^
  |
  = help: maybe it is overwritten before being read?
  = note: `#[warn(unused_assignments)]` on by default
```

This seems correct: the _value_ is unused, even though the _variable_ is used. The initial assignment can be removed (as can the `mut` keyword).

Do you have suggestions for making this warning clearer?

@rustbot label +A-diagnostics +T-compiler -C-bug

**Comment by jruderman** [comment]
Others have reported similar confusion: #90503, #81829

**Comment by c410-f3r** [comment]
Sorry, my bad. I should have paid more attention about the unread **initial** value.

Closing as duplicated.
